src/cursor/common.rs

```diff
@@ -0,0 +1,194 @@
+use std::{
+    collections::VecDeque,
+    pin::Pin,
+    task::{Context, Poll},
+    time::Duration,
+};
+
+use bson::Document;
+use derivative::Derivative;
+use futures::{Future, Stream};
+
+use crate::{
+    error::{ErrorKind, Result},
+    options::StreamAddress,
+    Client,
+    Namespace,
+};
+
+/// An internal cursor that can be used in a variety of contexts depending on its `GetMoreProvider`.
+#[derive(Derivative)]
+#[derivative(Debug)]
+pub(super) struct GenericCursor<T: GetMoreProvider> {
```

Comment:
I'm not sure I fully understand this explanation, so I apologize for asking questions that might seem obvious. What exactly do you mean by "wrap" in "no way to wrap a cursor that doesn't own its own session"? What does "handle" refer to in steps 1 and 2 (the session itself, or something else)? If "handle" refers to the session, is there any issue with storing that in the `state` field that currently exists and is used to hold onto the executing future?

I think I'm having a bit of trouble visual exactly what the code looked like in steps 1 and 2, which makes it somewhat hard to understand the compiler errors that occurred and therefore how they need to be fixed. Is it possible for you to make self-contained examples that produce the same compiler errors as steps 1 and 2, or barring that, separate branches with the non-compiling code?

Reply:
"handle" was referring to the struct returned from `cursor.with_session(&mut session)`

The original idea I had was to implement the explicit/borrowed session based cursor first and have the implicit/owned session based cursor simply wrap it and pass a reference to its owned session as needed. 

My first attempt (1) looked something like this:
```rust
struct Cursor {
    session: ClientSession,
    wrapped_cursor: SessionCursor
}

impl Stream for Cursor {
    fn poll_next(...) -> Poll {
        Pin::new(self.wrapped_cursor.with_session(&mut self.session)).poll_next()
    }
}
```
This didn't work because as soon as `Cursor::poll_next` exited, the handle created within the body would be dropped, along with any running getMore future.

My next plan (2) was to persist the handle between calls to `poll_next` to prevent it from being dropped too early. The problem with this is that the future stored in the handle contains a reference to the session which is also stored in `Cursor`, so storing it or something that owned it on `Cursor` would be self-referential.

```rust
struct CursorHandle<'session, 'cursor> {
    s: &'session mut ClientSession,
    c: &'cursor mut Cursor
}

struct Cursor {
    session: ClientSession,
    cursor: Cursor,
    handle: CursorHandle<???, ???>
}
```

So at this point, I figured that wrapping the borrowed cursor was simply not going to work (without something like `rental`, which seemed very ugly) since it'd require that the `Cursor` borrow from itself. The remaining alternatives were to either just duplicate all the cursor code for each cursor or find a way to abstract the parts that differ. Right now, these cursors only differ in their ownership of the session they're using, and that difference only surfaces in the form of the getMore future. So I defined a trait that abstracts over the type of the get more future and its lifetimes, and made a generic cursor that could operate over this abstraction so long as it was able to start getMores and yield results.

> is there any issue with storing that in the state field that currently exists and is used to hold onto the executing future?

When the session is owned, the future can be `BoxFuture<'static>`, since it isn't tied to the lifetime of the session. When the session is borrowed, the future must be `BoxFuture<'session>`. This is the "surfacing" of the differences I referred to in the previous paragraph, and it is the reason why the common cursor type needs to be generic over the future. It also needs to know how to create such a future, so I defined a trait that requires both.

Reply:
Okay, I think I understand now. Thanks for all the explanation.

Rather than having two separate cursor types, would it be possible for us to use something like [`Cow`](https://doc.rust-lang.org/std/borrow/enum.Cow.html) to abstract over whether the cursor is owned or borrowed?

Reply:
Never mind, I was thinking that the session types implemented Clone, but they don't, so `Cow` wouldn't work

Reply:
I think part of my confusion was due to the fact that we only ever use `GenericCursor` in a single place with the same generic parameter, so I didn't quite understand why we needed more than one type. Am I correct in understanding that the borrowed version would be returned from the explicit sessions API?

Reply:
I think overall I'd prefer if the names for things reflected the expected usage of them rather than just the mechanics of them. Naming the cursors `ExplicitSessionCursor` and `ImplicitSessionCursor` and the executor method `execute_cursor_operation` makes it easier to understand and probably would have caused a little less confusion for me at the beginning; I think most of my confusion was due to a lot of new types being introduced and my not really seeing how they all fit together. If you're okay with renaming things like that, then making those changes and then rebasing the changed from #169 should put this in a state where I can do one final pass over it to make sure I understand it all.

Reply:
> I think part of my confusion was due to the fact that we only ever use `GenericCursor` in a single place with the same generic parameter, so I didn't quite understand why we needed more than one type. Am I correct in understanding that the borrowed version would be returned from the explicit sessions API?

The borrowed version will be used with the explicit sessions API. Most of the internal work for that is actually already done, including using `GenericCursor` in a different place with a different generic parameter (see [`cursor/session.rs`](https://github.com/patrickfreed/mongo-rust-driver/blob/RUST-289/implicit-sessions/src/cursor/session.rs#L42)).

>I think overall I'd prefer if the names for things reflected the expected usage of them rather than just the mechanics of them.

This is totally fine with me, and in fact I chose the names to be as literal as possible with the expectation that we'd come up with better ones through review. I'm happy to change the names of any types introduced in this PR.

>If you're okay with renaming things like that, then making those changes and then rebasing the changed from #169 should put this in a state where I can do one final pass over it to make sure I understand it all.

Done

Reply:
Awesome, thanks for making the changes and for bearing with me during my confusion about all this